host/frontend/host-orchestrator/clients.go

```diff
@@ -0,0 +1,141 @@
+	"math/rand"
+	"sync"
+	"time"
+)
+
+// Implements the client interface using a websocket connection
+type WsClient struct {
+	ws *JsonWs
+}
+
+func NewWsClient(ws *JsonWs) *WsClient {
+	return &WsClient{ws: ws}
+}
+
+// From IClient
+func (c *WsClient) Send(msg interface{}) bool {
+	return c.ws.Send(msg)
+}
+func (c *WsClient) OnDeviceDisconnected() {
+	// Do nothing
+}
+
+// Implements the client interface using a polled connection
+type PolledClient struct {
+	// The polled connection id
+	id       string
+	device   *Device
+	messages []interface{}
+	// Synchronizes access to the messages list
+	msgMtx sync.Mutex
+	// The id given to this client by the device
+	clientId int
+	// A reference to the polled set to clean up when the device disconnects
+	set *PolledSet
+}
+
+// From IClient
+func (c *PolledClient) Send(msg interface{}) bool {
+	c.msgMtx.Lock()
+	defer c.msgMtx.Unlock()
+	c.messages = append(c.messages, msg)
+	return true
+}
+func (c *PolledClient) OnDeviceDisconnected() {
+	// This object is useless when the device disconnects, release all resourcess associated to it
+	c.set.Destroy(c.id)
+}
+
+// Sends a message to the device
+func (c *PolledClient) ToDevice(msg interface{}) bool {
+	return c.device.Send(msg)
+}
+
+// Gets count messages from the device, skipping the first start messages
+func (c *PolledClient) GetMessages(start int, count int) []interface{} {
+	c.msgMtx.Lock()
+	defer c.msgMtx.Unlock()
+	if count < 0 {
+		count = len(c.messages) - start
+	}
+	end := start + count
+	if end > len(c.messages) {
+		end = len(c.messages)
+	}
+	// Safe to return slice because c.messages is never shrunk
+	return c.messages[start:end]
+}
+
+func (c *PolledClient) Id() string {
+	return c.id
+}
+
+func (c *PolledClient) ClientId() int {
+	return c.clientId
+}
+
+// Basically a map of polled clients by id. This is necessary because polled
+// connections are not handled in an exclusive thread like those based on websockets.
+type PolledSet struct {
+	connections map[string]*PolledClient
+	mapMtx      sync.Mutex
+}
+
+func (s *PolledSet) NewConnection(d *Device) *PolledClient {
+	s.mapMtx.Lock()
+	defer s.mapMtx.Unlock()
+	for {
+		id := randStr(64)
+		if _, ok := s.connections[id]; !ok {
+			conn := &PolledClient{id: id, device: d, messages: make([]interface{}, 0), set: s}
+			s.connections[id] = conn
+			conn.clientId = d.Register(conn)
+			return conn
+		}
+	}
+}
+
+func (s *PolledSet) GetConnection(id string) *PolledClient {
+	s.mapMtx.Lock()
+	defer s.mapMtx.Unlock()
+	return s.connections[id]
+}
+
+func (s *PolledSet) Destroy(id string) {
+	s.mapMtx.Lock()
+	defer s.mapMtx.Unlock()
+	delete(s.connections, id)
+}
+
+func NewPolledSet() *PolledSet {
+	rand.Seed(time.Now().UnixNano())
+	return &PolledSet{connections: make(map[string]*PolledClient)}
+}
+
+// The available characters for the connection ids
+const runes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+func randStr(l int) string {
+	s := make([]byte, l)
```

Comment:
Can this use strings.Builder?

https://pkg.go.dev/strings#example-Builder

Reply:
Done